### PR TITLE
Generate: Bring certgen inline with cilium-cli

### PR DESCRIPTION
**Pull request by soggiest**
Fixes: [#300](https://github.com/cilium/cilium-cli/issues/300)

This PR brings the certgen tool in line with how `cilium-cli` [generates](https://github.com/cilium/cilium-cli/blob/master/defaults/defaults.go#L16) CA [certificates](https://github.com/cilium/cilium-cli/blob/master/internal/certs/certs.go#L92-L96). These two tools being out of sync with how they generate causes issues like those found in  [#300](https://github.com/cilium/cilium-cli/issues/300). Since `cilium-cli` appears to be the standard Cilium is using moving froward, based on the documentation, this PR proposes a simple solution to ensure the `cilium-cli` will work as expected when Cilium is installed via Helm.

I elected not to change the names of the variables defined in `defaults.go` so as to avoid changing the actual code as much as possible. Since both the [Hubble](https://github.com/cilium/certgen/blob/master/cmd/certgen.go#L185-L198) and [Certmesh](https://github.com/cilium/certgen/blob/master/cmd/certgen.go#L264-L277) CA generation code sections validate the existence of the CA secret there is no concern that the CA will be overwritten.

Signed-off-by: Nicholas Lane <nicklaneovi@gmail.com>

### CODE REVIEW

File: internal/defaults/defaults.go
Comment (kaworu) [inline internal/defaults/defaults.go:29]: The config map has been deprecated in v1.10 and v1.11 will remove it completely (see https://github.com/cilium/cilium/pull/17294). Instead of renaming it, we should consider removing it instead. cc @gandro

**Review by kaworu — CHANGES_REQUESTED** [review]
Hi @soggiest and thank you for the PR, it is a welcome change!

Some steps needs to be taken for the transition to happen smoothly:

1. The way certgen create CA secrets before this patch (i.e. `hubble-ca-secret` and `clustermesh-apiserver-ca-cert`) is consistent with [Cilium's default `helm` method](https://github.com/cilium/cilium/blob/af6b7953dc5d65033abb3d202c613a4532c214d0/install/kubernetes/cilium/values.yaml#L1710-L1716) and we should keep it that way. So once this PR is merged, a follow-up PR in Cilium updating the certgen version along with bringing the same changes to the Helm method should be opened. Let us know if you're interested in tackling this as well, I'll be happy to provide any pointers needed!
2. There must be a Cilium version (ideally v1.12) deprecating (without removing) both  `hubble-ca-secret` and `clustermesh-apiserver-ca-cert` in favor of using `cilium-ca` instead. In other words, this PR should be reworked in order to generate all three secrets, and once v1.12 is released we will be able to remove the deprecated ones (namely `hubble-ca-secret` and `clustermesh-apiserver-ca-cert`) from both the Helm and certgen method.

File: internal/defaults/defaults.go
Comment (gandro) [reply internal/defaults/defaults.go:29]: Agreed that we should remove this. Doesn't have to be part of this PR though, unless if it helps with the transition.

**Review by kaworu — CHANGES_REQUESTED** [review]
Great work @soggiest :pray: I have a minor nit and question but overall LGTM.

File: cmd/certgen.go
Comment (gandro) [reply cmd/certgen.go:80]: Yes. I would even be fine if we removed them fully. I don't think `certgen` has any uses outside of the Cilium Helm charts and therefore if the Cilium Helm charts will remove those flags, I don't see a reason to keep them around.

File: internal/defaults/defaults.go
Comment (kaworu) [reply internal/defaults/defaults.go]: Opened https://github.com/cilium/certgen/pull/43 to deprecate the hubble CA ConfigMap.

File: cmd/certgen.go
Comment (kaworu) [reply cmd/certgen.go:80]: sounds good to me, should we also [remove instead of deprecating the hubble ca configmap](https://github.com/cilium/certgen/pull/43) as it has already been removed from Cilium?

Comment (soggiest) [reply cmd/certgen.go:80]: My understanding based on a previous comment is that the deprecation of the `HubbleCA*` and `ClustermeshApiserverCA*` flags needs to be deprecated first in a Cilium release and then the logic for generating these CAs will be removed in a subsequent release. Based on this conversation it sounds like we are in favor of "ripping off the bandage" and removing the flags and logic all at once. If that's the case I'm fine with removing the flags and logic as part of this PR.

Comment (gandro) [reply cmd/certgen.go:80]: Yes, I am very much in favor off just removing the flags and associated logic without prior deprecation. You may do it in this PR, but we can also do it in a separate PR later, whatever seems easier.

Comment (soggiest) [reply cmd/certgen.go:80]: Burn it down

File: cmd/certgen.go
Comment (soggiest) [reply cmd/certgen.go:171]: Good catch! I've incorporated this change.

File: cmd/certgen.go
Comment (soggiest) [reply cmd/certgen.go]: I've removed these flags and the generation logic in my latest commit. There might be some redundancy with #43, if we want to merge that PR first I can rebase on top of it.

Comment (kaworu) [reply cmd/certgen.go]: Let's merge this one before as #43 will be easier to rebase.

Comment (soggiest) [reply cmd/certgen.go]: If everything looks good would you mind approving it?

**Review by gandro — APPROVED** [review]
Thanks, the code looks good to me. Also thanks a ton for cleaning up the old flag.

One open question however is how users are expected to upgrade to this new mode. With the current implementation, we will simply ignore any existing CA and create a new one. For Hubble Relay, will cause disruptions until all Cilium pods have been upgraded to accept the new CA.

This is not a problem specific to certgen, the `method=helm` based approach will suffer from the same problem. We should at document this caveat once we apply these changes to the Cilium Helm chart. If users have been using only one of the two CAs, they could likely rename the secret and update without much disruptions.

**Review by kaworu — CHANGES_REQUESTED** [review]
Thanks @soggiest. If found a small typo but other than that LGTM!

**Comment by soggiest** [comment]
@kaworu Good catch on that, I've incorporated your change directlyr.

**Review by kaworu — APPROVED** [review]

**Comment by gandro** [comment]
Looks ready to merge. I guess we can cut a release after that so we can pull in the new version in the Cilium helm charts?

**Comment by kaworu** [comment]
@soggiest @gandro I created [this issue](https://github.com/cilium/cilium/issues/18278) in cilium/cilium to keep track of what to do, let's move the discussion there.

**Pull request merged**
